### PR TITLE
TariBot key server

### PR DESCRIPTION
**Pull request by Jasonvdb**
- Signs message and requests UTXO drop from key server
- Fixed timezone conversion
- Reusable user feedback dialog class
- Fixes unit tests clashing with existing device wallet

### CODE REVIEW

File: MobileWallet/Screens/Home/HomeViewController.swift [hunk 1]
```diff
@@ -178,6 +178,35 @@ class HomeViewController: UIViewController, FloatingPanelControllerDelegate, Tra
...
+                        actionTitle: NSLocalizedString("Send Tari", comment: "Home view testnet airdrop"),
+                        onAction: {
+                            guard let _ = self else { return }
+                            self?.onSend()
+                        }
+                    )
+                }
+            }) { (error) in
+                DispatchQueue.main.async { [weak self] in
+                    guard let _ = self else { return }
```
Comment (lupugabriel) [inline MobileWallet/Screens/Home/HomeViewController.swift]: no need for guard let _ = self if you do not use self.

File: MobileWallet/Screens/Home/HomeViewController.swift [hunk 2]
```diff
@@ -178,6 +178,35 @@ class HomeViewController: UIViewController, FloatingPanelControllerDelegate, Tra
...
+                    UserFeedback.shared.callToAction(
+                        title: NSLocalizedString("You got some Tari!", comment: "Home view testnet airdrop"),
+                        description: NSLocalizedString("TariBot has just sent you some Tari. To give the wallet a quick test, try sending TariBot back some Tari to see how it works.", comment: "Home view testnet airdrop"),
+                        cancelTitle: NSLocalizedString("I’ll try it later", comment: "Home view testnet airdrop"),
+                        actionTitle: NSLocalizedString("Send Tari", comment: "Home view testnet airdrop"),
+                        onAction: {
+                            guard let _ = self else { return }
+                            self?.onSend()
```
Comment (lupugabriel) [inline MobileWallet/Screens/Home/HomeViewController.swift]: why not using guard let  self = self else { return }
self.onSend()?

File: MobileWallet/TariLib/Wrappers/Transactions/PendingInboundTransactions.swift [hunk 3]
```diff
@@ -84,7 +84,7 @@ class PendingInboundTransactions: TransactionsProtocol {
             }
         }
 
-        let sortedList = list.sorted(by: { $0.localDate.0?.compare($1.localDate.0!) == .orderedDescending })
+        let sortedList = list.sorted(by: { $0.date.0?.compare($1.date.0!) == .orderedDescending })
```
Comment (lupugabriel) [inline MobileWallet/TariLib/Wrappers/Transactions/PendingInboundTransactions.swift]: Its not ok to use !. maybe you check this also on UnitTesting?

File: MobileWallet/TariLib/Wrappers/Transactions/PendingOutboundTransactions.swift [hunk 4]
```diff
@@ -79,7 +79,7 @@ class PendingOutboundTransactions: TransactionsProtocol {
             }
         }
 
-        let sortedList = list.sorted(by: { $0.localDate.0?.compare($1.localDate.0!) == .orderedDescending })
+        let sortedList = list.sorted(by: { $0.date.0?.compare($1.date.0!) == .orderedDescending })
```
Comment (lupugabriel) [inline MobileWallet/TariLib/Wrappers/Transactions/PendingOutboundTransactions.swift]: same here. You used !

File: MobileWallet/TariLib/Wrappers/Utils/TestnetKeyServer.swift [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+    private let url: URL
+    private let wallet: Wallet
+
+    init(wallet: Wallet) throws {
+        let (publicKey, publicKeyError) = wallet.publicKey
+        guard publicKeyError == nil else {
+            throw publicKeyError!
+        }
+
+        let (publicKeyHex, hexError) = publicKey!.hex
```
Comment (lupugabriel) [inline MobileWallet/TariLib/Wrappers/Utils/TestnetKeyServer.swift]: here i think you used ! becaus you checkwd with guard if there is an error. So i guess it's fine.

File: MobileWallet/TariLib/Wrappers/Wallet.swift [hunk 6]
```diff
@@ -358,23 +359,40 @@ class Wallet {
...
+        guard publicKeyError == nil else {
+            throw publicKeyError!
+        }
+
+        let splitResult = result.components(separatedBy: "|")
+        guard splitResult.count == 2 else {
+            throw WalletErrors.invalidSignatureAndNonceString
+        }
+
+        return Signature(hex: splitResult[0], nonce: splitResult[1], message: message, publicKey: walletPubKey!)
```
Comment (lupugabriel) [inline MobileWallet/TariLib/Wrappers/Wallet.swift]: same here. probably you checked for error. there is no error then it's safe to use !

File: MobileWallet/TariLib/Wrappers/Transactions/PendingInboundTransactions.swift (hunk 3 above)
Comment (Jasonvdb) [reply MobileWallet/TariLib/Wrappers/Transactions/PendingInboundTransactions.swift]: I wasn't sure how to handle this sorted callback with optional dates. I don't think a unit test can help as the Tari lib can cause an issue I'm not aware of. I'll make an issue to come back to and find a better way to handle optional dates.

Comment (Jasonvdb) [reply MobileWallet/TariLib/Wrappers/Transactions/PendingInboundTransactions.swift]: https://github.com/tari-project/wallet-ios/issues/new

File: MobileWallet/TariLib/Wrappers/Transactions/PendingOutboundTransactions.swift (hunk 4 above)
Comment (Jasonvdb) [reply MobileWallet/TariLib/Wrappers/Transactions/PendingOutboundTransactions.swift]: https://github.com/tari-project/wallet-ios/issues/new

File: MobileWallet/TariLib/Wrappers/Utils/TestnetKeyServer.swift (hunk 5 above)
Comment (Jasonvdb) [reply MobileWallet/TariLib/Wrappers/Utils/TestnetKeyServer.swift]: Yes it's fine, with TariLib if the error is nil, the result will never be nil

File: MobileWallet/TariLib/Wrappers/Wallet.swift (hunk 6 above)
Comment (Jasonvdb) [reply MobileWallet/TariLib/Wrappers/Wallet.swift]: Yes same with TariLib if the error is nil, the result will never be nil

**Pull request merged**